### DISCUSSION TITLE
How can I install Phalcon on Debian?

### DISCUSSION THREAD

**Discussion by s-ohnishi** [phalcon/cphalcon#15537] [Q&A]
I'm trying to install Phalcon 4 on Debian 10.
Debian10 is prepared by Docker.
( php:7.4-fpm-buster )

I installed what was written as needed and added the repository.
```
# curl -s https://packagecloud.io/install/repositories/phalcon/stable/script.deb.sh | bash
# apt update
```

I will check it just in case.
```
# apt search phalcon
Sorting... Done
Full Text Search... Done
php7.3-phalcon/buster 4.1.2-1221+php7.3 amd64
  (none)

php7.3-phalcon-dbgsym/buster 4.1.2-1221+php7.3 amd64
  (none)

php7.4-phalcon/buster 4.1.2-1221+php7.4 amd64
  (none)

php7.4-phalcon-dbgsym/buster 4.1.2-1221+php7.4 amd64
  (none)
```

And when I try to install it, it looks like this.
```
# apt install php7.4-phalcon
Reading package lists... Done
Building dependency tree
Reading state information... Done
Package php7.4-phalcon is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source

E: Package 'php7.4-phalcon' has no installation candidate
```

What am I doing wrong?

**Comment by quasipickle** [comment]
Buster only has packages for PHP 7.3: https://wiki.debian.org/PHP#Available_versions .  I'm guessing the fact you have 7.4 is because the Docker image (not sure what the proper noun is here) your using is based on the unstable version.

Does `php -v` report version 7.4 or version 7.3?

As for why `apt search` finds a package, but you can't install it - maybe this will help?: https://askubuntu.com/questions/391184/but-is-referred-to-by-another-package-finding-that-package (Ubuntu is based on Debian so a lot of the answers are backwards compatible).  A lot of the answers seem to be "run `apt-update`" - which you've already done.

**Reply by s-ohnishi** [reply]
> Does php -v report version 7.4 or version 7.3?
Of course it is 7.4.
Because, as shown above, I am using the official PHP image php: 7.4-fpm-buster.

```
# php -v
PHP 7.4.20 (cli) (built: Jun  3 2021 21:55:37) ( NTS )
Copyright (c) The PHP Group
Zend Engine v3.4.0, Copyright (c) Zend Technologies
```

The page you told me didn't work as you expected.
(Because update has already been executed)

But reading that page I had an idea, so I looked it up.
Then
/etc/apt/preferences.d/no-debian-php
File exists
```
Package: php *
Pin: release *
Pin-Priority: -1
```
Was written.
Since it is an image issued by PHP official, it may be a consideration so that other PHP will not be inadvertently mixed in.
However, it also rejects php7.4-phalcon.

So I changed the look of this file so that it wasn't used by the apt system and the response changed.
```
# mv /etc/apt/preferences.d/no-debian-php /etc/apt/preferences.d/no.txt
# apt install php7.4-phalcon
Reading package lists... Done
Building dependency tree
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 php7.4-phalcon : Depends: phpapi-20190902 but it is not installable
N: Ignoring file 'no.txt' in directory '/etc/apt/preferences.d/' as it has an invalid filename extension
E: Unable to correct problems, you have held broken packages.
```

What should I do after this?
'phpapi-20190902' is not supported by Baster (it was phpapi-20180731 in Buster) and the official PHP image does not include that version either.
Is this a problem caused by using the official PHP image with Docker?
Is there an option to use PHP 7.3 just before EOL?
Isn't it possible that 'phpapi-20190902', on which phalcon depends, can be 'phpapi-20180731' or older?

**Reply by s-ohnishi** [reply]
I gave up adding Phalcon to the official PHP image using apt.
As I wrote above, it is basically prohibited to use the package named php *.
Therefore, although it takes time, I decided to compile it by myself.

`# curl -LO https://github.com/phalcon/cphalcon/archive/v4.1.2.tar.gz`
`# tar xzf v4.1.2.tar.gz`
`# docker-php-ext-install -j$(nproc) ${PWD}/cphalcon-4.1.2/build/php7/64bits`
or 
`# pecl install phalcon-4.1.2`
`# echo "extension=phalcon.so" > /etc/php/7.4/mods-available/phalcon.ini`
`# phpenmod -v 7.4 -s ALL phalcon`

Both methods give a warning, but for now it seems that there is no problem in using it.
`/root/cphalcon-4.1.2/build/php7/64bits/phalcon.zep.c:11022:2: warning: writing 2 bytes into a region of size 1 [-Wstringop-overflow=]`
`11022 |  memcpy(Z_STRVAL_P(result) + offset + op1_len + Z_STRLEN_P(op2) + op3_len + Z_STRLEN_P(op4) + op5_len + op6_len + Z_STRLEN_P(op7), op8, op8_len);`
(I'm not sure if it really doesn't matter)

**Reply by Jeckerson** [reply]
Warning messages can be ignored.

If you want to install via OS package manager, please use:
https://packagecloud.io/phalcon/stable

**Reply by s-ohnishi** [reply]
> Warning messages can be ignored.

That's right, the warning message can be ignored.
But as an engineer, it doesn't feel good, right?

> If you want to install via OS package manager, please use:
 https://packagecloud.io/phalcon/stable

Using packagecloud may be available if you installed PHP individually on plain Debian, but based on the official PHP image, it wasn't available as I wrote in the first post.

And now that time has passed and Buster has changed to Bullseye, there seems to be no package for Bullseye in packagecloud.
(Of course, the lack of a package for Bullseye is not directly related to this thread, because it was a post from the time I was using Buster.)
